services/debugger/cache/badgercache/badgercache.go

```diff
@@ -0,0 +1,188 @@
+package embeddedcache
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"path"
+	"sync"
+	"time"
+
+	"github.com/dgraph-io/badger/v3"
+	"github.com/dgraph-io/badger/v3/options"
+	"github.com/rudderlabs/rudder-server/config"
+	"github.com/rudderlabs/rudder-server/rruntime"
+	"github.com/rudderlabs/rudder-server/utils/logger"
+	"github.com/rudderlabs/rudder-server/utils/misc"
+)
+
+/*
+loadCacheConfig sets the properties of the cache after reading it from the config file.
+This gives a feature of hot readability as well.
+*/
+func (e *EmbeddedCache[E]) loadCacheConfig() {
+	config.RegisterDurationConfigVariable(30, &e.cleanupFreq, true, time.Second, "LiveEvent.cache.clearFreq") // default clearFreq is 15 seconds
+	config.RegisterIntConfigVariable(100, &e.limiter, true, 1, "LiveEvent.cache.limiter")
+	config.RegisterDurationConfigVariable(5, &e.ticker, false, time.Minute, "LiveEvent.cache.GCTime")
+	config.RegisterDurationConfigVariable(15, &e.queryTimeout, false, time.Second, "LiveEvent.cache.queryTimeout")
+	config.RegisterIntConfigVariable(3, &e.retries, false, 1, "LiveEvent.cache.retries")
+}
+
+/*
+EmbeddedCache is an in-memory cache. Each key-value pair stored in this cache have a TTL and one goroutine removes the
+key-value pair form the cache which is older than TTL time.
+*/
+type EmbeddedCache[E any] struct {
+	limiter      int
+	retries      int
+	path         string
+	Origin       string
+	done         chan struct{}
+	closed       chan struct{}
+	ticker       time.Duration
+	queryTimeout time.Duration
+	cleanupFreq  time.Duration // TTL time on badgerDB
+	once         sync.Once
+	db           *badger.DB
+	Logger       logger.Logger
+}
+
+type badgerLogger struct {
+	logger.Logger
+}
+
+func (badgerLogger) Errorf(format string, a ...interface{}) {
+	_, _ = fmt.Fprintf(os.Stderr, format, a...)
+}
+
+func (badgerLogger) Warningf(format string, a ...interface{}) {
+	_, _ = fmt.Fprintf(os.Stderr, format, a...)
+}
+
+// Update writes the entries into badger db with a TTL
+func (e *EmbeddedCache[E]) Update(key string, value E) error {
+	e.Init()
+	txn := e.db.NewTransaction(true)
+	setBadgerEntry := func(ctx context.Context) error {
+		res := e.Read(key)
+		if len(res) >= e.limiter {
+			res = res[len(res)-e.limiter+1:]
+		}
+		res = append(res, value)
+		data, err := json.Marshal(res)
+		if err != nil {
+			return err
+		}
+		entry := badger.NewEntry([]byte(key), data).WithTTL(e.cleanupFreq)
+		if err := txn.SetEntry(entry); err != nil {
+			return err
+		}
+		return nil
+	}
+	err := misc.RetryWith(context.TODO(), e.queryTimeout, e.retries, setBadgerEntry)
+	if err != nil {
+		return err
+	}
+	err = txn.Commit()
```

Comment:
we should retry here if the error is `ErrConflict`

Reply:
Done the same